plugins/workspace-backpack/src/index.js

```diff
@@ -438,43 +432,44 @@ export class Backpack {
    *     cleaned of all unnecessary attributes.
    */
   addItem(item) {
-    if (this.contents_.indexOf(item) !== -1) {
-      return;
-    }
+    this.addItems([item]);
+  }
 
-    this.contents_.unshift(item);
-    while (this.contents_.length > this.maxItems_) {
-      this.contents_.pop();
-    }
-    // TODO: Fire UI event for Backpack content change.
+  /**
+   * Adds multiple items to the backpack.
+   * @param {!Array<string>} items The backpack contents to add.
+   */
+  addItems(items) {
+    this.contents_.unshift(...this.filterDuplicates_(items));
+    Blockly.Events.fire(new BackpackChange(this.workspace_.id));
   }
 
   /**
    * Sets backpack contents XML.
    * @param {!Array<string>} contents The new backpack contents.
    */
   setContents(contents) {
-    this.contents_ = [...contents];
-    while (this.contents_.length > this.maxItems_) {
-      this.contents_.pop();
-    }
-    // TODO: Fire UI event for Backpack content change.
+    this.contents_ = this.filterDuplicates_(contents);
```

Comment:
I think this one might still have a problem.
Block A is already in the backpack.
Call setContents(A, B, C)
results in the backpack containing B, C

(These would be good unit test candidates if you're planning to add them)

Reply:
That is a good one. I have a branch for tests but I was having trouble getting them to run because I need to mock more things to make it work.

I've added a setting this.contents_ to empty array before this line, which I think fixes the issue.